File: pkg/remediation/safety_checker.go
package remediation

import (
	"context"
	"fmt"

	"github.com/namansh70747/aura-k8s/pkg/k8s"
	"github.com/namansh70747/aura-k8s/pkg/utils"
	"github.com/sirupsen/logrus"
	corev1 "k8s.io/api/core/v1"
	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
	"k8s.io/apimachinery/pkg/labels"
)

// SafetyCheckResult represents the result of a safety check
type SafetyCheckResult string

const (
	SafetyCheckPass SafetyCheckResult = "pass"
	SafetyCheckWarn SafetyCheckResult = "warn"
	SafetyCheckFail SafetyCheckResult = "fail"
	SafetyCheckSkip SafetyCheckResult = "skip"
)

// SafetyCheck represents a single safety check
type SafetyCheck struct {
	Name    string
	Result  SafetyCheckResult
	Message string
	Details map[string]interface{}
}

// RemediationSafetyChecker performs comprehensive safety checks
type RemediationSafetyChecker struct {
	k8sClient *k8s.Client
	logger    *logrus.Logger
}

// NewRemediationSafetyChecker creates a new safety checker
func NewRemediationSafetyChecker(k8sClient *k8s.Client) *RemediationSafetyChecker {
	return &RemediationSafetyChecker{
		k8sClient: k8sClient,
		logger:    utils.Log,
	}
}

// Validate performs all safety checks for an action
func (s *RemediationSafetyChecker) Validate(
	ctx context.Context,
	action RemediationAction,
	pod *corev1.Pod,
) (bool, []SafetyCheck) {
	checks := []SafetyCheck{}

	// Run all checks
	checks = append(checks, s.checkPDBViolations(ctx, action, pod))
	checks = append(checks, s.checkResourceAvailability(ctx, action, pod))
	checks = append(checks, s.checkDependencies(ctx, action, pod))
	checks = append(checks, s.checkQuotaLimits(ctx, action, pod))
	checks = append(checks, s.checkPermissions(ctx, action, pod))
	checks = append(checks, s.checkRollbackCapability(ctx, action, pod))

	// Check if any critical checks failed
	hasFailures := false
	for _, check := range checks {
		if check.Result == SafetyCheckFail {
			hasFailures = true
			break
		}
	}

	return !hasFailures, checks
}

// checkPDBViolations checks if action would violate Pod Disruption Budget
func (s *RemediationSafetyChecker) checkPDBViolations(
	ctx context.Context,
	action RemediationAction,
	pod *corev1.Pod,
) SafetyCheck {
	// Check if action involves pod deletion/eviction
	if action.Type == "pod" && (action.Operation == "restart" || action.Operation == "delete" || action.Operation == "evict") {
		// Check for PDB
		pdbs, err := s.k8sClient.Clientset().PolicyV1().PodDisruptionBudgets(pod.Namespace).List(ctx, metav1.ListOptions{})
		if err != nil {
			return SafetyCheck{
				Name:    "pdb_check",
				Result:  SafetyCheckWarn,
				Message: fmt.Sprintf("Could not check PDB: %v", err),
				Details: map[string]interface{}{"error": err.Error()},
			}
		}

		// Check if pod matches any PDB
		for _, pdb := range pdbs.Items {
			selector, err := metav1.LabelSelectorAsSelector(pdb.Spec.Selector)
			if err != nil {
				continue
			}

			if selector.Matches(labels.Set(pod.Labels)) {
				// Check if eviction would violate PDB
				// DisruptionsAllowed is an int32 field in policy/v1
				disruptionsAllowed := pdb.Status.DisruptionsAllowed
				if disruptionsAllowed == 0 {
					return SafetyCheck{
						Name:    "pdb_check",
						Result:  SafetyCheckFail,
						Message: fmt.Sprintf("Action would violate PDB %s (0 disruptions allowed)", pdb.Name),
						Details: map[string]interface{}{
							"pdb_name":            pdb.Name,
							"disruptions_allowed": 0,
						},
					}
				}
			}
		}
	}

	return SafetyCheck{
		Name:    "pdb_check",
		Result:  SafetyCheckPass,
		Message: "PDB check passed",
		Details: map[string]interface{}{},
	}
}

// checkResourceAvailability checks if resources are available
func (s *RemediationSafetyChecker) checkResourceAvailability(
	_ context.Context,
	action RemediationAction,
	_ *corev1.Pod,
) SafetyCheck {
	if action.Operation == "scale" || action.Operation == "scale_up" {
		replicas, _ := action.Parameters["replicas"].(float64)
		if replicas > 10 {
			return SafetyCheck{
				Name:    "resource_availability",
				Result:  SafetyCheckWarn,
				Message: fmt.Sprintf("Large scale operation (%v replicas) - verify cluster capacity", replicas),
				Details: map[string]interface{}{"replicas": replicas},
			}
		}
	}

	return SafetyCheck{
		Name:    "resource_availability",
		Result:  SafetyCheckPass,
		Message: "Resource availability check passed",
		Details: map[string]interface{}{},
	}
}

// checkDependencies checks action dependencies
func (s *RemediationSafetyChecker) checkDependencies(
	_ context.Context,
	action RemediationAction,
	_ *corev1.Pod,
) SafetyCheck {
	if action.Type == "deployment" {
		if action.Target == "" || action.Target == "deployment" {
			return SafetyCheck{
				Name:    "dependencies",
				Result:  SafetyCheckFail,
				Message: "Deployment target not specified",
				Details: map[string]interface{}{},
			}
		}
	}

	return SafetyCheck{
		Name:    "dependencies",
		Result:  SafetyCheckPass,
		Message: "Dependency check passed",
		Details: map[string]interface{}{},
	}
}

// checkQuotaLimits checks resource quota limits
func (s *RemediationSafetyChecker) checkQuotaLimits(
	_ context.Context,
	action RemediationAction,
	_ *corev1.Pod,
) SafetyCheck {
	operation := action.Operation
	parameters := action.Parameters

	if operation == "increase_memory" || operation == "increase_cpu" {
		factor, ok := parameters["factor"].(float64)
		if ok && factor > 3.0 {
			return SafetyCheck{
				Name:    "quota_limits",
				Result:  SafetyCheckWarn,
				Message: fmt.Sprintf("Large resource increase (factor %.2f) may exceed quota", factor),
				Details: map[string]interface{}{"factor": factor},
			}
		}
	}

	return SafetyCheck{
		Name:    "quota_limits",
		Result:  SafetyCheckPass,
		Message: "Quota check passed",
		Details: map[string]interface{}{},
	}
}

// checkPermissions checks if user has required permissions
func (s *RemediationSafetyChecker) checkPermissions(
	_ context.Context,
	_ RemediationAction,
	_ *corev1.Pod,
) SafetyCheck {
	// Permission checks would be implemented here
	// For now, assume permissions are valid
	return SafetyCheck{
		Name:    "permissions",
		Result:  SafetyCheckPass,
		Message: "Permission check passed",
		Details: map[string]interface{}{},
	}
}

// checkRollbackCapability checks if action can be rolled back
func (s *RemediationSafetyChecker) checkRollbackCapability(
	_ context.Context,
	action RemediationAction,
	_ *corev1.Pod,
) SafetyCheck {
	nonRollbackable := []string{"delete", "force_delete"}
	for _, op := range nonRollbackable {
		if action.Operation == op {
			return SafetyCheck{
				Name:    "rollback_capability",
				Result:  SafetyCheckWarn,
				Message: fmt.Sprintf("Operation %s cannot be rolled back", action.Operation),
				Details: map[string]interface{}{"rollbackable": false},
			}
		}
	}

	return SafetyCheck{
		Name:    "rollback_capability",
		Result:  SafetyCheckPass,
		Message: "Rollback capability check passed",
		Details: map[string]interface{}{"rollbackable": true},
	}
}

// PreCheck performs pre-execution safety checks
func (s *RemediationSafetyChecker) PreCheck(
	ctx context.Context,
	action RemediationAction,
	pod *corev1.Pod,
) (bool, []string) {
	passed, checks := s.Validate(ctx, action, pod)

	warnings := []string{}
	for _, check := range checks {
		if check.Result == SafetyCheckWarn {
			warnings = append(warnings, check.Message)
		}
	}

	return passed, warnings
}

// PostCheck performs post-execution validation
func (s *RemediationSafetyChecker) PostCheck(
	ctx context.Context,
	action RemediationAction,
	pod *corev1.Pod,
	result map[string]interface{},
) (bool, []string) {
	success, ok := result["success"].(bool)
	if !ok || !success {
		return false, []string{fmt.Sprintf("Action %s failed", action.Operation)}
	}

	return true, []string{}
}

File: mcp/safety_checker.py
"""
Safety & Validation Framework
Comprehensive safety checks before remediation execution
"""

from typing import Dict, List, Optional, Any, Tuple
from dataclasses import dataclass
from enum import Enum
import logging

logger = logging.getLogger(__name__)


class SafetyCheckResult(str, Enum):
    PASS = "pass"
    WARN = "warn"
    FAIL = "fail"
    SKIP = "skip"


@dataclass
class SafetyCheck:
    """Result of a safety check"""
    name: str
    result: SafetyCheckResult
    message: str
    details: Dict[str, Any] = None


class RemediationSafetyChecker:
    """Comprehensive safety checker for remediations"""
    
    def __init__(self, k8s_tools=None):
        self.k8s_tools = k8s_tools
        self.checks = []
    
    def validate(
        self,
        action: Dict[str, Any],
        context: Dict[str, Any]
    ) -> Tuple[bool, List[SafetyCheck]]:
        """Validate action with all safety checks"""
        results = []
        
        # Run all checks
        results.append(self.check_pdb_violations(action, context))
        results.append(self.check_resource_availability(action, context))
        results.append(self.check_dependencies(action, context))
        results.append(self.check_quota_limits(action, context))
        results.append(self.check_permissions(action, context))
        results.append(self.check_rollback_capability(action, context))
        
        # Check if any critical checks failed
        has_failures = any(
            check.result == SafetyCheckResult.FAIL
            for check in results
        )
        
        return not has_failures, results
    
    def check_pdb_violations(
        self,
        action: Dict[str, Any],
        context: Dict[str, Any]
    ) -> SafetyCheck:
        """Check Pod Disruption Budget violations"""
        # This would check if action would violate PDB
        # For now, return pass (would need K8s API access)
        return SafetyCheck(
            name="pdb_check",
            result=SafetyCheckResult.PASS,
            message="PDB check passed (not implemented)",
            details={}
        )
    
    def check_resource_availability(
        self,
        action: Dict[str, Any],
        context: Dict[str, Any]
    ) -> SafetyCheck:
        """Check if resources are available"""
        action_type = action.get("type", "")
        operation = action.get("operation", "")
        
        # Check if scaling up would exceed cluster capacity
        if operation in ["scale", "scale_up"]:
            replicas = action.get("parameters", {}).get("replicas", 1)
            # Would check cluster capacity here
            return SafetyCheck(
                name="resource_availability",
                result=SafetyCheckResult.WARN,
                message=f"Scaling by {replicas} replicas - verify cluster capacity",
                details={"replicas": replicas}
            )
        
        return SafetyCheck(
            name="resource_availability",
            result=SafetyCheckResult.PASS,
            message="Resource availability check passed",
            details={}
        )
    
    def check_dependencies(
        self,
        action: Dict[str, Any],
        context: Dict[str, Any]
    ) -> SafetyCheck:
        """Check action dependencies"""
        # Check if action depends on other resources
        action_type = action.get("type", "")
        
        if action_type == "deployment":
            deployment_name = action.get("target", "")
            if not deployment_name or deployment_name == "deployment":
                return SafetyCheck(
                    name="dependencies",
                    result=SafetyCheckResult.FAIL,
                    message="Deployment target not specified",
                    details={}
                )
        
        return SafetyCheck(
            name="dependencies",
            result=SafetyCheckResult.PASS,
            message="Dependency check passed",
            details={}
        )
    
    def check_quota_limits(
        self,
        action: Dict[str, Any],
        context: Dict[str, Any]
    ) -> SafetyCheck:
        """Check resource quota limits"""
        operation = action.get("operation", "")
        parameters = action.get("parameters", {})
        
        # Check if resource increase would exceed quota
        if operation in ["increase_memory", "increase_cpu"]:
            factor = parameters.get("factor", 1.0)
            if factor > 3.0:
                return SafetyCheck(
                    name="quota_limits",
                    result=SafetyCheckResult.WARN,
                    message=f"Large resource increase (factor {factor}) may exceed quota",
                    details={"factor": factor}
                )
        
        return SafetyCheck(
            name="quota_limits",
            result=SafetyCheckResult.PASS,
            message="Quota check passed",
            details={}
        )
    
    def check_permissions(
        self,
        action: Dict[str, Any],
        context: Dict[str, Any]
    ) -> SafetyCheck:
        """Check if user has required permissions"""
        # Would check RBAC permissions here
        return SafetyCheck(
            name="permissions",
            result=SafetyCheckResult.PASS,
            message="Permission check passed (not implemented)",
            details={}
        )
    
    def check_rollback_capability(
        self,
        action: Dict[str, Any],
        context: Dict[str, Any]
    ) -> SafetyCheck:
        """Check if action can be rolled back"""
        operation = action.get("operation", "")
        
        non_rollbackable = ["delete", "force_delete"]
        if operation in non_rollbackable:
            return SafetyCheck(
                name="rollback_capability",
                result=SafetyCheckResult.WARN,
                message=f"Operation {operation} cannot be rolled back",
                details={"rollbackable": False}
            )
        
        return SafetyCheck(
            name="rollback_capability",
            result=SafetyCheckResult.PASS,
            message="Rollback capability check passed",
            details={"rollbackable": True}
        )
    
    def pre_check(
        self,
        action: Dict[str, Any],
        context: Dict[str, Any]
    ) -> Tuple[bool, List[str]]:
        """Pre-execution safety checks"""
        passed, checks = self.validate(action, context)
        
        warnings = [
            check.message
            for check in checks
            if check.result == SafetyCheckResult.WARN
        ]
        
        return passed, warnings
    
    def post_check(
        self,
        action: Dict[str, Any],
        context: Dict[str, Any],
        result: Dict[str, Any]
    ) -> Tuple[bool, List[str]]:
        """Post-execution validation checks"""
        # Verify action completed successfully
        success = result.get("success", False)
        
        if not success:
            return False, [f"Action {action.get('operation')} failed"]
        
        # Additional post-checks could go here
        return True, []



